vkubelet/api/exec.go

```diff
@@ -26,30 +37,121 @@ func PodExecHandlerFunc(backend remotecommand.Executor) http.HandlerFunc {
 		q := req.URL.Query()
 		command := q["command"]
 
-		var stdin, stdout, stderr, tty bool
-		if q.Get("stdin") == "true" {
-			stdin = true
-		}
-		if q.Get("stdout") == "true" {
-			stdout = true
-		}
-		if q.Get("stderr") == "true" {
-			stderr = true
-		}
-		if q.Get("tty") == "true" {
-			tty = true
-		}
-		
-		streamOpts := &remotecommand.Options{
-			Stdin:  stdin,
-			Stdout: stdout,
-			Stderr: stderr,
-			TTY:    tty,
+		streamOpts, err := getExecOptions(req)
+		if err != nil {
+			return strongerrors.InvalidArgument(err)
 		}
 
 		idleTimeout := time.Second * 30
 		streamCreationTimeout := time.Second * 30
 
-		remotecommand.ServeExec(w, req, backend, fmt.Sprintf("%s-%s", namespace, pod), "", container, command, streamOpts, idleTimeout, streamCreationTimeout, supportedStreamProtocols)
+		ctx, cancel := context.WithCancel(context.TODO())
+		defer cancel()
+
+		exec := &containerExecContext{ctx: ctx, b: backend, pod: pod, namespace: namespace, container: container}
+		remotecommand.ServeExec(w, req, exec, "", "", container, command, streamOpts, idleTimeout, streamCreationTimeout, supportedStreamProtocols)
```

Comment:
I would ultimately like to have not used this function, however all the stream setup is handled in here and most of it is not exported.
There's just a lot to handle in here and I did not want to duplicate it and have potential divergence in the future.
This includes support for different websocket protocol versions as well as SPDY.